### PR TITLE
VP-2704: Add expressions for the rules

### PR DESCRIPTION
**Pull request by yecli**
Started expression tree

### CODE REVIEW

File: src/VirtoCommerce.CatalogModule.Core/Model/DynamicAssociations/DynamicAssociationEvaluationContextExtensions.cs [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+        {
+            categoryIds = categoryIds.Where(x => x != null).ToArray();
+            var productArray = products as CatalogProduct[] ?? products.ToArray();
+            return categoryIds.Any() ? productArray.Where(x => ProductInCategories(x, categoryIds)) : productArray;
+        }
+
+        public static bool ProductInCategories(this CatalogProduct product, ICollection<string> categoryIds)
+        {
+            var result = categoryIds.Contains(product.CategoryId, StringComparer.OrdinalIgnoreCase);
+            if (!result && !product.Outlines.IsNullOrEmpty())
```
Comment (pushnitsa) [inline src/VirtoCommerce.CatalogModule.Core/Model/DynamicAssociations/DynamicAssociationEvaluationContextExtensions.cs:54]: Add empty lines before and after if braces

File: src/VirtoCommerce.CatalogModule.Core/Model/DynamicAssociations/DynamicAssociationEvaluationContextExtensions.cs [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+
+        public static bool ProductInProducts(this CatalogProduct product, IEnumerable<string> productIds)
+        {
+            return productIds.Contains(product.Id, StringComparer.OrdinalIgnoreCase);
+        }
+
+        public static IEnumerable<CatalogProduct> WithPropertyValues(this IEnumerable<CatalogProduct> products, Dictionary<string, string> propertyValues)
+        {
+            var productArray = products as CatalogProduct[] ?? products.ToArray();
+            return propertyValues.Any() ? productArray.Where(x => x.ProductHasPropertyValues(propertyValues)) : productArray;
```
Comment (pushnitsa) [inline src/VirtoCommerce.CatalogModule.Core/Model/DynamicAssociations/DynamicAssociationEvaluationContextExtensions.cs:69]: Add empty line before return

File: src/VirtoCommerce.CatalogModule.Core/Model/DynamicAssociations/DynamicAssociationEvaluationContextExtensions.cs [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+        public static IEnumerable<CatalogProduct> ExcludeProducts(this IEnumerable<CatalogProduct> products, string[] productIds)
+        {
+            var retval = products.Where(x => !ProductInProducts(x, productIds));
+            return retval;
+        }
+
+        public static IEnumerable<CatalogProduct> InCategories(this IEnumerable<CatalogProduct> products, string[] categoryIds)
+        {
+            categoryIds = categoryIds.Where(x => x != null).ToArray();
+            var productArray = products as CatalogProduct[] ?? products.ToArray();
```
Comment (pushnitsa) [inline src/VirtoCommerce.CatalogModule.Core/Model/DynamicAssociations/DynamicAssociationEvaluationContextExtensions.cs:47]: Why do you made cast here?

File: src/VirtoCommerce.CatalogModule.Core/Model/DynamicAssociations/DynamicAssociationEvaluationContextExtensions.cs [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+        public static bool AreItemPropertyValuesEqual(this DynamicAssociationEvaluationContext context, Dictionary<string, string> propertyValues)
+        {
+            var result = context.Products.WithPropertyValues(propertyValues).Any();
+
+            return result;
+        }
+
+        #endregion DynamicAssociationEvaluationContext extensions
+
+        #region CatalogProduct extensions
```
[the remark this answers is not in the document]
Comment (yecli) [reply src/VirtoCommerce.CatalogModule.Core/Model/DynamicAssociations/DynamicAssociationEvaluationContextExtensions.cs:30]: Imho they are not the problem.

File: src/VirtoCommerce.CatalogModule.Core/Model/DynamicAssociations/Conditions/ConditionCategoryIs.cs [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+    public class ConditionCategoryIs : ConditionTree
+    {
+        public ICollection<string> ExcludingCategoryIds { get; set; } = new List<string>();
+        public ICollection<string> ExcludingProductIds { get; set; } = new List<string>();
+
+        public string[] CategoryIds { get; set; }
+        public string[] CategoryNames { get; set; }
+
+        /// <summary>
+        /// ((PromotionEvaluationContext)x).IsItemInCategory(CategoryId, ExcludingCategoryIds, ExcludingProductIds)
```
[the remark this answers is not in the document]
Comment (yecli) [reply src/VirtoCommerce.CatalogModule.Core/Model/DynamicAssociations/Conditions/ConditionCategoryIs.cs:17]: Just the existing code copy, will remove.

File: src/VirtoCommerce.CatalogModule.Core/Model/DynamicAssociations/DynamicAssociationEvaluationContextExtensions.cs (hunk 2 above)
Comment (yecli) [reply src/VirtoCommerce.CatalogModule.Core/Model/DynamicAssociations/DynamicAssociationEvaluationContextExtensions.cs:69]: Just the existing code copy, will change.

File: src/VirtoCommerce.CatalogModule.Core/Model/DynamicAssociations/DynamicAssociationEvaluationContextExtensions.cs (hunk 1 above)
Comment (yecli) [reply src/VirtoCommerce.CatalogModule.Core/Model/DynamicAssociations/DynamicAssociationEvaluationContextExtensions.cs:54]: Just the existing code copy, will change.

File: src/VirtoCommerce.CatalogModule.Core/Model/DynamicAssociations/DynamicAssociationEvaluationContextExtensions.cs (hunk 4 above)
[the remark this answers is not in the document]
Comment (pushnitsa) [reply src/VirtoCommerce.CatalogModule.Core/Model/DynamicAssociations/DynamicAssociationEvaluationContextExtensions.cs:30]: But we have a coding convention, where doesn't recommended using regions https://github.com/VirtoCommerce/styleguide/blob/master/csharp.md

File: src/VirtoCommerce.CatalogModule.Core/Model/DynamicAssociations/DynamicAssociationEvaluationContextExtensions.cs (hunk 3 above)
Comment (yecli) [reply src/VirtoCommerce.CatalogModule.Core/Model/DynamicAssociations/DynamicAssociationEvaluationContextExtensions.cs:47]: Just the existing code copy, will change.

File: src/VirtoCommerce.CatalogModule.Core/Model/DynamicAssociations/DynamicAssociationEvaluationContextExtensions.cs (hunk 4 above)
[the remark this answers is not in the document]
Comment (yecli) [reply src/VirtoCommerce.CatalogModule.Core/Model/DynamicAssociations/DynamicAssociationEvaluationContextExtensions.cs:30]: Indeed, there are such rules in our coding conventions, regardless we have a lot of code with the regions.
I do not see problems in regions at all personally (they are suitable sometimes), but if it is our code style - will remove.

**Pull request merged**